<H2>Importing Libraries</H2>

In [1]:
#Importing Libraries

import pandas as pd
import numpy as np
import os

In [2]:
#Create path

path =r'C:\Users\supri\Desktop\Supriya\DataAnalysis\Project\Data Immersion\Achievement 1-2\Instacart_Basket_Analysis'

<H3>Importing Dataframes</H3>

In [4]:
#Importing data

df_orders_products_combined = pd.read_pickle(os.path.join(path,'02 Data', 'Prepared Data','orders_products_combined.pkl'))

df_prods= pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data','products_checked.csv'),index_col= False)

In [5]:
#Check output
df_orders_products_combined.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id eval_set  order_number  \
0             0           0   2539329        1    prior             1   
1             0           0   2539329        1    prior             1   
2             0           0   2539329        1    prior             1   
3             0           0   2539329        1    prior             1   
4             0           0   2539329        1    prior             1   

   orders_days_of_week  order_hour_of_day  days_since_prior_order  \
0                    2                  8                     NaN   
1                    2                  8                     NaN   
2                    2                  8                     NaN   
3                    2                  8                     NaN   
4                    2                  8                     NaN   

   first_time_customers  product_id  add_to_cart_order  reordered _merge  
0                  True         196                  1          0   both  
1                  True       14084                  2          0   both  
2                  True       12427                  3          0   both  
3                  True       26088                  4          0   both  
4                  True       26405                  5          0   both

In [6]:
#Check output
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [7]:
#Check shape of the dataframe
df_orders_products_combined.shape

(32434489, 14)

In [8]:
#Check shape of the dataframe
df_prods.shape

(49672, 6)

<H3>Merging the DataFrames</H3>

In [9]:
#Clean the dataframes before merging by dropping the unnecessary columns 

df_orders_products_combined = df_orders_products_combined.drop(['Unnamed: 0', '_merge'], axis =1)

In [10]:
df_orders_products_combined

Unnamed: 0.1  order_id  user_id eval_set  order_number  \
0                    0   2539329        1    prior             1   
1                    0   2539329        1    prior             1   
2                    0   2539329        1    prior             1   
3                    0   2539329        1    prior             1   
4                    0   2539329        1    prior             1   
...                ...       ...      ...      ...           ...   
32434484       3421081   2977660   206209    prior            13   
32434485       3421081   2977660   206209    prior            13   
32434486       3421081   2977660   206209    prior            13   
32434487       3421081   2977660   206209    prior            13   
32434488       3421081   2977660   206209    prior            13   

          orders_days_of_week  order_hour_of_day  days_since_prior_order  \
0                           2                  8                     NaN   
1                           2                  8                     NaN   
2                           2                  8                     NaN   
3                           2                  8                     NaN   
4                           2                  8                     NaN   
...                       ...                ...                     ...   
32434484                    1                 12                     7.0   
32434485                    1                 12                     7.0   
32434486                    1                 12                     7.0   
32434487                    1                 12                     7.0   
32434488                    1                 12                     7.0   

          first_time_customers  product_id  add_to_cart_order  reordered  
0                         True         196                  1          0  
1                         True       14084                  2          0  
2                         True       12427                  3          0  
3                         True       26088                  4          0  
4                         True       26405                  5          0  
...                        ...         ...                ...        ...  
32434484                 False       14197                  5          1  
32434485                 False       38730                  6          0  
32434486                 False       31477                  7          0  
32434487                 False        6567                  8          0  
32434488                 False       22920                  9          0  

[32434489 rows x 12 columns]

In [11]:
df_prods = df_prods.drop(['Unnamed: 0'], axis =1)

In [12]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49667       49684          Vodka, Triple Distilled, Twist of Vanilla   
49668       49685                 En Croute Roast Hazelnut Cranberry   
49669       49686                                   Artisan Baguette   
49670       49687         Smartblend Healthy Metabolism Dry Cat Food   
49671       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49667       124              5     5.3  
49668        42              1     3.1  
49669       112              3     7.8  
49670        41              8     4.7  
49671        73             11    13.5  

[49672 rows x 5 columns]

<H4>Observation -The 'product_id' column is present in both datasets allowing us to merge both the datasets together.We information that exists in both dataframes, therefore, the 'inner' merge will be a suitable fit. </H4>

In [13]:
# Test merge without overwriting with merge flag 'match'

pd.merge(df_orders_products_combined , df_prods , on = ['product_id'] , how = 'inner', indicator = 'match' )

Unnamed: 0.1  order_id  user_id eval_set  order_number  \
0                    0   2539329        1    prior             1   
1                    0   2539329        1    prior             1   
2                    0   2539329        1    prior             1   
3                    0   2539329        1    prior             1   
4                    0   2539329        1    prior             1   
...                ...       ...      ...      ...           ...   
32404854       3421081   2977660   206209    prior            13   
32404855       3421081   2977660   206209    prior            13   
32404856       3421081   2977660   206209    prior            13   
32404857       3421081   2977660   206209    prior            13   
32404858       3421081   2977660   206209    prior            13   

          orders_days_of_week  order_hour_of_day  days_since_prior_order  \
0                           2                  8                     NaN   
1                           2                  8                     NaN   
2                           2                  8                     NaN   
3                           2                  8                     NaN   
4                           2                  8                     NaN   
...                       ...                ...                     ...   
32404854                    1                 12                     7.0   
32404855                    1                 12                     7.0   
32404856                    1                 12                     7.0   
32404857                    1                 12                     7.0   
32404858                    1                 12                     7.0   

          first_time_customers  product_id  add_to_cart_order  reordered  \
0                         True         196                  1          0   
1                         True       14084                  2          0   
2                         True       12427                  3          0   
3                         True       26088                  4          0   
4                         True       26405                  5          0   
...                        ...         ...                ...        ...   
32404854                 False       14197                  5          1   
32404855                 False       38730                  6          0   
32404856                 False       31477                  7          0   
32404857                 False        6567                  8          0   
32404858                 False       22920                  9          0   

                                     product_name  aisle_id  department_id  \
0                                            Soda        77              7   
1         Organic Unsweetened Vanilla Almond Milk        91             16   
2                             Original Beef Jerky        23             19   
3                      Aged White Cheddar Popcorn        23             19   
4                XL Pick-A-Size Paper Towel Rolls        54             17   
...                                           ...       ...            ...   
32404854                             Tomato Paste         9              9   
32404855          Brownie Crunch High Protein Bar         3             19   
32404856    High Protein Bar Chunky Peanut Butter         3             19   
32404857      Chocolate Peanut Butter Protein Bar         3             19   
32404858      Roasted & Salted Shelled Pistachios       117             19   

          prices match  
0            9.0  both  
1           12.5  both  
2            4.4  both  
3            4.7  both  
4            1.0  both  
...          ...   ...  
32404854     5.6  both  
32404855     5.9  both  
32404856     4.2  both  
32404857     4.9  both  
32404858     1.7  both  

[32404859 rows x 17 columns]

<H5>The test merge indicates that the merge with inner join still gives all the 32404859 rows so goign ahead with the actual merge </H5>

In [14]:
# Actual merge of dataframes 'df_orders_products_combined' and 'df_prods'

ords_prods_merge = df_orders_products_combined.merge(df_prods , on = ['product_id'] , indicator = 'match' )

In [15]:
ords_prods_merge.head()

Unnamed: 0.1  order_id  user_id eval_set  order_number  \
0             0   2539329        1    prior             1   
1             0   2539329        1    prior             1   
2             0   2539329        1    prior             1   
3             0   2539329        1    prior             1   
4             0   2539329        1    prior             1   

   orders_days_of_week  order_hour_of_day  days_since_prior_order  \
0                    2                  8                     NaN   
1                    2                  8                     NaN   
2                    2                  8                     NaN   
3                    2                  8                     NaN   
4                    2                  8                     NaN   

   first_time_customers  product_id  add_to_cart_order  reordered  \
0                  True         196                  1          0   
1                  True       14084                  2          0   
2                  True       12427                  3          0   
3                  True       26088                  4          0   
4                  True       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  match  
0  both  
1  both  
2  both  
3  both  
4  both

In [16]:
#Check output
ords_prods_merge.shape

(32404859, 17)

In [18]:
#Check value counts on Match column
ords_prods_merge['match'].value_counts()

match
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

<H5>The newly merged dataset has 32404859 rows matching both of the original dataset.</H5>

In [19]:
#Exporting the data in pickle format

ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge.pkl'))